Address diffs v2.12.1 to v3

chengzhuzhang opened the pull request:
## Description

<!--
  Please include a summary of the change and which issue is fixed.
  Please also include relevant motivation and context.
  List any dependencies that are required for this change.
-->

- Closes #906

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] My changes generate no new warnings
- [ ] Any dependent changes have been merged and published in downstream modules

If applicable:

- [ ] New and existing unit tests pass with my changes (locally and CI/CD build)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have noted that this is a breaking change for a major release (fix or feature that would cause existing functionality to not work as expected)

tomvothecoder commented:
Regression testing performed in #903, related to [comment](https://github.com/E3SM-Project/e3sm_diags/pull/903#issuecomment-2537406898).

Copying here:

Here the [regression notebook](https://github.com/E3SM-Project/e3sm_diags/blob/tests/894-regression-test/auxiliary_tools/cdat_regression_testing/894-regression-test/regression_nc.ipynb) that compares latest `main` vs. `v2.12.1`.

- Matching number of `.nc` files produced (590)
- 501/590 are matching
- 4/590 are mismatching (`nan` positions)
- 85/590 are not equal

For the 85/590 are not equal to `rtol=1e-4`. These variables will need further investigation.
```python
['/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/COREv2_Flux/COREv2_Flux-PminusE-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/CRU_IPCC/CRU-TREFHT-ANN-land_60S90N_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/CRU_IPCC/CRU-TREFHT-ANN-land_60S90N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/Cloud MISR/MISRCOSP-CLDTOT_TAU1.3_9.4_MISR-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/Cloud MISR/MISRCOSP-CLDTOT_TAU1.3_MISR-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-NET_FLUX_SRF-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-OMEGA-200-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-OMEGA-500-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-OMEGA-850-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-TREFHT-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-TREFHT-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-TREFHT-ANN-land_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/ERA5/ERA5-U-850-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/GPCP_OAFLux/GPCP_OAFLux-PminusE-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-NET_FLUX_SRF-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-OMEGA-200-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-OMEGA-500-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-OMEGA-850-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFHT-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFHT-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFHT-ANN-land_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFMNAV-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFMNAV-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFMXAV-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/MERRA2/MERRA2-TREFMXAV-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/SST_CL_HadISST/HadISST_CL-SST-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/SST_HadISST/HadISST-SST-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/SST_PD_HadISST/HadISST_PD-SST-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/lat_lon/SST_PI_HadISST/HadISST_PI-SST-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/meridional_mean_2d/MERRA2/MERRA2-OMEGA-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/CRU_IPCC/CRU-TREFHT-ANN-polar_N_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/CRU_IPCC/CRU-TREFHT-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/ERA5/ERA5-TREFHT-ANN-polar_N_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/ERA5/ERA5-TREFHT-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/ERA5/ERA5-TREFHT-ANN-polar_S_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/ERA5/ERA5-TREFHT-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/ERA5/ERA5-U-850-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/ERA5/ERA5-U-850-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFHT-ANN-polar_N_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFHT-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFHT-ANN-polar_S_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFHT-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMNAV-ANN-polar_N_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMNAV-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMNAV-ANN-polar_S_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMNAV-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMXAV-ANN-polar_N_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMXAV-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMXAV-ANN-polar_S_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/MERRA2/MERRA2-TREFMXAV-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/SST_CL_HadISST/HadISST_CL-SST-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/SST_CL_HadISST/HadISST_CL-SST-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/SST_PD_HadISST/HadISST_PD-SST-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/SST_PD_HadISST/HadISST_PD-SST-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/SST_PI_HadISST/HadISST_PI-SST-ANN-polar_N_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/polar/SST_PI_HadISST/HadISST_PI-SST-ANN-polar_S_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_2d/ERA5/ERA5-OMEGA-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_2d/MERRA2/MERRA2-OMEGA-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_2d_stratosphere/ERA5/ERA5-H2OLNZ-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_2d_stratosphere/ERA5/ERA5-H2OLNZ-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_2d_stratosphere/MERRA2/MERRA2-H2OLNZ-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_2d_stratosphere/MERRA2/MERRA2-H2OLNZ-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/ERA5/ERA5-TREFHT-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/ERA5/ERA5-TREFHT-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/MERRA2/MERRA2-TREFHT-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/MERRA2/MERRA2-TREFHT-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/MERRA2/MERRA2-TREFMNAV-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/MERRA2/MERRA2-TREFMNAV-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/MERRA2/MERRA2-TREFMXAV-ANN-global_ref.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/MERRA2/MERRA2-TREFMXAV-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/SST_CL_HadISST/HadISST_CL-SST-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/SST_PD_HadISST/HadISST_PD-SST-ANN-global_test.nc',
 '/global/cfs/cdirs/e3sm/www/e3sm_diags/complete_run/24-12-09-main/zonal_mean_xy/SST_PI_HadISST/HadISST_PI-SST-ANN-global_test.nc']
```

tomvothecoder commented:
I will pull my regression testing notebooks over from PR #903 to this PR and work on debugging in this branch with you.

That way #903 can be dedicated to adapting the regression tests.

chengzhuzhang commented:
thank you @tomvothecoder for summarizing the cases to investigate. It look like I'm running into environment issue while debugging with vscode https://github.com/E3SM-Project/e3sm_diags/issues/906#issuecomment-2537209291. Let me know if the env works for you when debugging.

tomvothecoder commented:
> thank you @tomvothecoder for summarizing the cases to investigate. It look like I'm running into environment issue while debugging with vscode [#906 (comment)](https://github.com/E3SM-Project/e3sm_diags/issues/906#issuecomment-2537209291). Let me know if the env works for you when debugging.

I just replied [here](https://github.com/E3SM-Project/e3sm_diags/issues/906#issuecomment-2540001982). The dev env works fine for me.

tomvothecoder commented:
I think I found the root cause of the large diffs here and in your comment [here](https://github.com/E3SM-Project/e3sm_diags/pull/880#issuecomment-2539886503), specifically with `TREFHT` and ANN.

### `v2.12.1` code uses `genutil.udunits`
In `v2.12.1`, the `convert_units()` function for derived vars includes a portion of logic that calls `genutil.udunits`:
https://github.com/E3SM-Project/e3sm_diags/blob/ca41b0e5d913610c88410928951f1ed11c75663f/e3sm_diags/derivations/acme.py#L93-L97

### Latest `main` uses `cf_units.Unit` with new implementation logic
I replaced `genutil.udnits` with `cf_units.Unit`. It looks like the conversion is incorrect for `TREFHT` between `K` to `C`. 
https://github.com/E3SM-Project/e3sm_diags/blob/df102713adde55fdf05fc8eb534a4375842194d4/e3sm_diags/derivations/utils.py#L85-L93

The implementation might not be able to handle all variables and/or the logic is incorrect, which leads to the massive diff found in your comment [here](https://github.com/E3SM-Project/e3sm_diags/pull/880#issuecomment-2539886503).


### Next steps

Investigate the `cf_units.Unit` code block in `convert_units()`. Fix as needed.

tomvothecoder commented:
> Investigate the `cf_units.Unit` code block in `convert_units()`. Fix as needed.

I just pushed [`ee009c1` (#907)](https://github.com/E3SM-Project/e3sm_diags/pull/907/commits/ee009c1bc6195b19009e01072431f278a804778b) to address this issue.

It fixes `ERA5-TREFHT-ANN-global`. Links are below.

- PR Plot: https://portal.nersc.gov/project/e3sm/cdat-migration-fy24/907-address-diffs/lat_lon/ERA5/ERA5-TREFHT-ANN-global.png
- Previous EAMxx plot: https://portal.nersc.gov/cfs/e3sm/zhang40/tests/eamxx/eamxx_decadal_1104/lat_lon/ERA5/ERA5-TREFHT-ANN-land.png

I will perform another regression test with the complete run to compare against `v2.12.1`.

tomvothecoder reviewed (commented):
@chengzhuzhang FYI if we are still experiencing slowness or hanging with a complete run in #880, I will carry this fix over to that PR as well and re-run.

For cases where deriving a climo variable occurs, this fix will first subset the dataset with the source variables and load it into memory before performing the actual derivation. 

This fix is required because the `convert_units()` function will try to load the variables into memory using the wrong scheduler by calling `.values` instead of `.load(scheduler="sync"`), which causes hanging/crashing _(code below)_.

https://github.com/E3SM-Project/e3sm_diags/blob/3ea542234a99fffdc81b2592ced6e4410fd91f73/e3sm_diags/derivations/utils.py#L85-L89

chengzhuzhang commented:
@tomvothecoder hey, Tom. Just a heads-up.  I think #880 is in good shape and ready to merge.
I'm happy to test the scripts from #880 again when this PR is finalized. Let me know if anything I can help for this PR...

tomvothecoder commented:
> @tomvothecoder hey, Tom. Just a heads-up. I think #880 is in good shape and ready to merge. 

Great, I'll perform a final review before merging. Thanks!

> I'm happy to test the scripts from #880 again when this PR is finalized. Let me know if anything I can help for this PR...

In this PR, I’m debugging the final variables that fail to generate for `arm_diags`. The issue stems from sub-optimal/incorrect chunk sizes in the native v3 datasets, causing the Xarray `.load(scheduler="sync")` call to hang due to excessive chunks.

Here’s a script from [`af046a2` (#907)](https://github.com/E3SM-Project/e3sm_diags/pull/907/commits/af046a29af90c1470ac5f4227e892fb0fabf731d) (pasted below for convenience) detailing the affected datasets and their chunk size issues. Any thoughts on why these chunk sizes were set this way?

```python
# %%
"""
Issue 1 - Sub-optimal `CLOUD` and `time_bnds chunking
  * Related dataset: "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/CLOUD_twpc3_200001_201412.nc"
  * Dataset shape is (time: 131400, bound: 2, lev: 80)
  * `CLOUD` variable has chunks of (1, 80), resulting in 131400 chunks in 2 graph layers. (very bad, slow loading)
  * `time_bnds` has chunks of (1, 2), resulting in 131400 chunks in 3 graph layers. (very bad, slow loading)
"""

import xcdat as xc

ds = xc.open_mfdataset(
    [
        "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/CLOUD_twpc3_200001_201412.nc"
    ]
)

print(ds.CLOUD.data)
#               Array	        Chunk
# Bytes	        40.10 MiB	    320 B
# Shape	        (131400, 80)	(1, 80)
# Dask graph	131400 chunks in 2 graph layers
# Data type	    float32 numpy.ndarray

print(ds.time_bnds.data)
# Array	Chunk
# Bytes	2.01 MiB	16 B
# Shape	(131400, 2)	(1, 2)
# Dask graph	131400 chunks in 3 graph layers
# Data type	object numpy.ndarray


# %%
"""
Issue 2 - Sub-optimal `time_bnds` chunking
  * Related dataset: "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/FLDS_sgpc1_200001_201412.nc"
  * Dataset shape is (time: 131400, bound: 2, lev: 80)
  * `FLDS` variable has chunks of (1019,), resulting in 129 in 2 graph layers (okay)
  * `time_bnds` has chunks of (1, 2), resulting in 131400 chunks in 3 graph layers (very bad, slow loading)
"""

ds2 = xc.open_mfdataset(
    [
        "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/FLDS_sgpc1_200001_201412.nc"
    ]
)

print(ds2.FLDS.data)
#               Array	Chunk
# Bytes	        513.28 kiB	3.98 kiB
# Shape	        (131400,)	(1019,)
# Dask graph	129 chunks in 2 graph layers
# Data type	    float32 numpy.ndarray

print(ds2.time_bnds.data)
#               Array	Chunk
# Bytes	        2.01 MiB	16 B
# Shape	        (131400, 2)	(1, 2)
# Dask graph	131400 chunks in 3 graph layers
# Data type	    object numpy.ndarray

# %%
"""
Issue 3 - Sub-optimal `time_bnds` chunking
  * Related dataset: "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/PRECT_sgpc1_200001_201412.nc"
  * Dataset shape is (time: 131400, bound: 2, lev: 80)
  * `PRECT` variable has chunks of (1019,), resulting in 129 in 2 graph layers (okay)
  * `time_bnds` has chunks of (1, 2), resulting in 131400 chunks in 3 graph layers (very bad, slow loading)
"""

ds3 = xc.open_mfdataset(
    [
        "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/PRECT_sgpc1_200001_201412.nc"
    ]
)

print(ds3.PRECT.data)
#               Array	Chunk
# Bytes	        513.28 kiB	3.98 kiB
# Shape	        (131400,)	(1019,)
# Dask graph	129 chunks in 2 graph layers
# Data type	    float32 numpy.ndarray

print(ds3.time_bnds.data)
#               Array	Chunk
# Bytes	        2.01 MiB	16 B
# Shape	        (131400, 2)	(1, 2)
# Dask graph	131400 chunks in 3 graph layers
# Data type	    object numpy.ndarray

# %%
"""
Issue 4 - Sub-optimal `num_a1` and `time_bnds` chunking
  * Related dataset: "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/num_a1_enac1_200001_201412.nc"
  * Dataset shape is (time: 131400, bound: 2, lev: 80)
  * `num_a1` variable has chunks of (1, 80), resulting in 131400 chunks in 2 graph layers. (very bad, slow loading)
  * `time_bnds` has chunks of (1, 2), resulting in 131400 chunks in 3 graph layers (very bad, slow loading)
"""

ds4 = xc.open_mfdataset(
    [
        "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/num_a1_enac1_200001_201412.nc"
    ]
)

print(ds4.num_a1.data)
#               Array	Chunk
# Bytes	        40.10 MiB	320 B
# Shape	        (131400, 80)	(1, 80)
# Dask graph	131400 chunks in 2 graph layers
# Data type	    float32 numpy.ndarray

print(ds4.time_bnds.data)
#               Array	Chunk
# Bytes	        2.01 MiB	16 B
# Shape	        (131400, 2)	(1, 2)
# Dask graph	131400 chunks in 3 graph layers
# Data type	    object numpy.ndarray


# %%
"""
Issue 5 - Sub-optimal `time_bnds` chunking
  * Related dataset: "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/PRECT_twpc3_200001_201412.nc"
  * Dataset shape is (time: 131400, bound: 2, lev: 80)
  * `PRECT` variable has chunks of (1019,), resulting in 129 in 2 graph layers (okay)
  * `time_bnds` has chunks of (1, 2), resulting in 131400 chunks in 3 graph layers (very bad, slow loading)
"""

ds5 = xc.open_mfdataset(
    [
        "/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/PRECT_twpc3_200001_201412.nc"
    ]
)

print(ds5.time_bnds.data)
#               Array	Chunk
# Bytes	        2.01 MiB	16 B
# Shape	        (131400, 2)	(1, 2)
# Dask graph	131400 chunks in 3 graph layers
# Data type	    object numpy.ndarray
```

chengzhuzhang commented:
@tomvothecoder I'm trying to understand the new emerging issues. I feel we haven't seen these when we merging the new code base. Another difference is that we are now testing a new dataset as well v3 as compared to v2 that is used for testing during refactor. 

It looks like firstly, changing `genutil.udunits` to `cf_units.Unit`, and then this change requires that instead of using `.value` we need `.load(scheduler="sync")`, and the latter caused wired chunking? Did I get the story right? I'm wondering if we will get same behavior if we re-test `arm-diags` with the v2 data we have been working with?

tomvothecoder commented:
> @tomvothecoder I'm trying to understand the new emerging issues. I feel we haven't seen these when we tested the refactor branch. Another difference is that we are now testing a new dataset as well v3 as compared to v2 that is used for testing during refactor.
> 

The root cause of the issue for `arm_diags` [above](https://github.com/E3SM-Project/e3sm_diags/pull/907#issuecomment-2578977046) is specific to the v3 data. The v3 data seems to have incorrect chunking at the file level, which causes Xarray to hang when trying to load into memory. This is not an issue with the codebase itself, as far as I'm aware.

 > It looks like firstly, `genutil.udunits` is changed to `cf_units.Unit`, and then this change requires that instead of using `.value` we need `.load(scheduler="sync")`, and the latter caused wired chunking? Did I get the story right? I'm wondering if we will get same behavior if we re-test `arm-diags` with the v2 data we have been working with?
> 
> I'm now testing with the v2 dataset here: /global/cfs/cdirs/e3sm/e3sm_diags/postprocessed_e3sm_v2_data_for_e3sm_diags/20221103.v2.LR.amip.NGD_v3atm.chrysalis/arm-diags-data/. The time record is longer in this v2 dataset which is from 1985-2014.

The units conversion topic is separate from the `arm_diags` issue.

chengzhuzhang commented:
I'm digging up some discussions regarding to performance of refactored `arm_diags` set: https://github.com/E3SM-Project/e3sm_diags/pull/842#issuecomment-2344166562. There are some nice performance benchmarks there.

chengzhuzhang commented:
> The root cause of the issue for `arm_diags` [above](https://github.com/E3SM-Project/e3sm_diags/pull/907#issuecomment-2578977046) is specific to the v3 data. The v3 data seems to have incorrect chunking at the file level (defined via `_Chunksizes`) for some variables and time_bnds, which causes Xarray to hang when trying to load into memory because there are way too many chunks. This is not an issue with the codebase itself, as far as I'm aware.

I checked the metadata for v2 vs v3. The difference is that v2 data has 2 x time steps of v3; v2 has 72 vertical levels and v3 has 80. Is it possible that in the original refactored code, data was loaded in without chunking, while in the new code base, data are loaded as dask arrray and chunked in a weird way?

tomvothecoder commented:
> I checked the metadata for v2 vs v3. The difference is that v2 data has 2 x time steps of v3; v2 has 72 vertical levels and v3 has 80. Is it possible that in the original refactored code, data was loaded in without chunking, while in the new code base, data are loaded as dask arrray and chunked in a weird way?

There are a few possibilities here (and any combination of them together):
1. We did regression testing on v2 data with `open_dataset()` which avoids Dask issues
2. We incorporated `open_mfdataset()` and `.load()` later on (e.g., #861) 
3. The v2 data does not have weird chunking like v3 data

As a workaround, I can add code to detect if there is only 1 file to open. If there is, we use `open_dataset()` instead of `open_mfdataset()`. It shouldn't affect anything downstream, but a regression test should still be performed with this change.

chengzhuzhang commented:
> 3\. The v2 data does not have weird chunking like v3 data

This is mysterious to me. Cause both v2 and v3 files were pre-processed with a same cdms-based script: [here](https://github.com/E3SM-Project/e3sm_diags/blob/main/examples/tutorials2024/generate_data_from_single_sites.py)), and [here](https://github.com/E3SM-Project/e3sm_diags/blob/main/model_data_preprocess/postprocessing_E3SM_data_for_single_sites.py). I don't know why the chunking would be different on file level. Another possibility is to rewrite this pre-processing script with xarray and create a new set of files to test with.

chengzhuzhang commented:
> This is mysterious to me, cause both v2 and v3 files were pre-processed with a same cdms-based script: [here](https://github.com/E3SM-Project/e3sm_diags/blob/main/examples/tutorials2024/generate_data_from_single_sites.py)), and [here](https://github.com/E3SM-Project/e3sm_diags/blob/main/model_data_preprocess/postprocessing_E3SM_data_for_single_sites.py). I don't know why the chunking would be different on file level. Another possibility is to rewrite this pre-processing script with xarray and create a new set of files to test with.

I did a little research on chunking. `ncks -m --hdn` is the NCO command to view chunking size. It does look like both version of data has similar chunking size. 
```
(e3sm_unified_1.10.0_login) chengzhu@perlmutter:login30:~/arm> ncks -m --hdn /global/cfs/cdirs/e3sm/e3sm_diags/postprocessed_e3sm_v2_data_for_e3sm_diags/20221103.v2.LR.amip.NGD_v3atm.chrysalis/arm-diags-data/CLOUD_twpc3_198501_201412.nc 
netcdf CLOUD_twpc3_198501_201412 {
  dimensions:
    bound = 2 ;
    lev = 72 ;
    time = UNLIMITED ; // (262800 currently)

  variables:
    float CLOUD(time,lev) ;
      CLOUD:mdims = 1 ;
      CLOUD:units = "1" ;
      CLOUD:long_name = "Cloud fraction" ;
      CLOUD:cell_methods = "time: point time: mean" ;
      CLOUD:basename = "CLOUD" ;
      CLOUD:missing_value = 1.e+20f ;
      CLOUD:_FillValue = 1.e+20f ;
      CLOUD:_Storage = "chunked" ;
      CLOUD:_ChunkSizes = 1, 72 ;
      CLOUD:_Endianness = "little" ;

    double P0 ;
      P0:name = "P0" ;
      P0:missing_value = 1.e+20 ;
      P0:_FillValue = 1.e+20 ;
      P0:_Endianness = "little" ;

    float PS(time) ;
      PS:units = "Pa" ;
      PS:long_name = "Surface pressure" ;
      PS:standard_name = "surface_air_pressure" ;
      PS:cell_methods = "time: point time: mean" ;
      PS:basename = "PS" ;
      PS:missing_value = 1.e+20f ;
      PS:_FillValue = 1.e+20f ;
      PS:_Storage = "chunked" ;
      PS:_ChunkSizes = 1023 ;
      PS:_Endianness = "little" ;
    double hyam(lev) ;
      hyam:long_name = "hybrid A coefficient at layer midpoints" ;
      hyam:missing_value = 1.e+20 ;
      hyam:_FillValue = 1.e+20 ;
      hyam:_Storage = "contiguous" ;
      hyam:_Endianness = "little" ;

    double hybm(lev) ;
      hybm:long_name = "hybrid B coefficient at layer midpoints" ;
      hybm:missing_value = 1.e+20 ;
      hybm:_FillValue = 1.e+20 ;
      hybm:_Storage = "contiguous" ;
      hybm:_Endianness = "little" ;

    double lat ;
      lat:long_name = "latitude" ;
      lat:units = "degrees_north" ;
      lat:missing_value = 1.e+20 ;
      lat:_FillValue = 1.e+20 ;
      lat:_Endianness = "little" ;

    double lev(lev) ;
      lev:axis = "Z" ;
      lev:units = "hPa" ;
      lev:long_name = "hybrid level at midpoints (1000*(A+B))" ;
      lev:positive = "down" ;
      lev:standard_name = "atmosphere_hybrid_sigma_pressure_coordinate" ;
      lev:formula_terms = "a: hyam b: hybm p0: P0 ps: PS" ;
      lev:realtopology = "linear" ;
      lev:_Storage = "contiguous" ;
      lev:_Endianness = "little" ;

    double lon ;
      lon:long_name = "longitude" ;
      lon:units = "degrees_east" ;
      lon:missing_value = 1.e+20 ;
      lon:_FillValue = 1.e+20 ;
      lon:_Endianness = "little" ;

    double time(time) ;
      time:bounds = "time_bnds" ;
      time:axis = "T" ;
      time:calendar = "noleap" ;
      time:units = "days since 1985-01-01 00:00:00" ;
      time:long_name = "time" ;
      time:cell_methods = "time: mean" ;
      time:realtopology = "linear" ;
      time:_Storage = "chunked" ;
      time:_ChunkSizes = 512 ;
      time:_Endianness = "little" ;

    double time_bnds(time,bound) ;
      time_bnds:_Storage = "chunked" ;
      time_bnds:_ChunkSizes = 1, 2 ;
      time_bnds:_Endianness = "little" ;
} // group /
```
and
```
(e3sm_unified_1.10.0_login) chengzhu@perlmutter:login30:~/arm> ncks -m --hdn /global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/CLOUD_twpc3_200001_201412.nc
netcdf CLOUD_twpc3_200001_201412 {
  dimensions:
    bound = 2 ;
    lev = 80 ;
    time = UNLIMITED ; // (131400 currently)

  variables:
    float CLOUD(time,lev) ;
      CLOUD:mdims = 1 ;
      CLOUD:units = "fraction" ;
      CLOUD:long_name = "Cloud fraction" ;
      CLOUD:cell_methods = "time: point" ;
      CLOUD:basename = "CLOUD" ;
      CLOUD:missing_value = 1.e+20f ;
      CLOUD:_FillValue = 1.e+20f ;
      CLOUD:_Storage = "chunked" ;
      CLOUD:_ChunkSizes = 1, 80 ;
      CLOUD:_Endianness = "little" ;

    double P0 ;
      P0:name = "P0" ;
      P0:missing_value = 1.e+20 ;
      P0:_FillValue = 1.e+20 ;
      P0:_Endianness = "little" ;

    float PS(time) ;
      PS:units = "Pa" ;
      PS:long_name = "Surface pressure" ;
      PS:standard_name = "surface_air_pressure" ;
      PS:cell_methods = "time: point" ;
      PS:basename = "PS" ;
      PS:missing_value = 1.e+20f ;
      PS:_FillValue = 1.e+20f ;
      PS:_Storage = "chunked" ;
      PS:_ChunkSizes = 1019 ;
      PS:_Endianness = "little" ;

    double hyam(lev) ;
      hyam:long_name = "hybrid A coefficient at layer midpoints" ;
      hyam:missing_value = 1.e+20 ;
      hyam:_FillValue = 1.e+20 ;
      hyam:_Storage = "contiguous" ;
      hyam:_Endianness = "little" ;

    double hybm(lev) ;
      hybm:long_name = "hybrid B coefficient at layer midpoints" ;
      hybm:missing_value = 1.e+20 ;
      hybm:_FillValue = 1.e+20 ;
      hybm:_Storage = "contiguous" ;
      hybm:_Endianness = "little" ;

    double lat ;
      lat:long_name = "latitude" ;
      lat:units = "degrees_north" ;
      lat:missing_value = 1.e+20 ;
      lat:_FillValue = 1.e+20 ;
      lat:_Endianness = "little" ;

    double lev(lev) ;
      lev:axis = "Z" ;
      lev:units = "hPa" ;
      lev:long_name = "hybrid level at midpoints (1000*(A+B))" ;
      lev:positive = "down" ;
      lev:standard_name = "atmosphere_hybrid_sigma_pressure_coordinate" ;
      lev:formula_terms = "a: hyam b: hybm p0: P0 ps: PS" ;
      lev:realtopology = "linear" ;
      lev:_Storage = "contiguous" ;
      lev:_Endianness = "little" ;

    double lon ;
      lon:long_name = "longitude" ;
      lon:units = "degrees_east" ;
      lon:missing_value = 1.e+20 ;
      lon:_FillValue = 1.e+20 ;
      lon:_Endianness = "little" ;

    double time(time) ;
      time:bounds = "time_bnds" ;
      time:axis = "T" ;
      time:calendar = "noleap" ;
      time:units = "days since 2000-01-01 00:00:00" ;
      time:long_name = "time" ;
      time:realtopology = "linear" ;
      time:_Storage = "chunked" ;
      time:_ChunkSizes = 512 ;
      time:_Endianness = "little" ;

    double time_bnds(time,bound) ;
      time_bnds:_Storage = "chunked" ;
      time_bnds:_ChunkSizes = 1, 2 ;
      time_bnds:_Endianness = "little" ;
} // group /

I think the chunking follows NCO's chunking policy, but for this particular small datasets, we probably don't need to chunk..I'm rewriting the pre-processing script using xarray/xcdat, to see if we can get more reasonable `.nc` files to start with.

tomvothecoder commented:
> I did a little research on chunking. `ncks -m --hdn` is the NCO command to view chunking size. It does look like both version of data has similar chunking size.

Thanks for confirming @chengzhuzhang. This leads me to believe that we ran the v2 complete run with `open_dataset()` rather than `open_mfdataset()` + `load()`, which explains why `arm_diags` worked fine.

> I think the chunking follows NCO's chunking policy, but for this particular small datasets, we probably don't need to chunk..I'm rewriting the pre-processing script using xarray/xcdat, to see if we can get more reasonable `.nc` files to start with.

Sounds good, thanks.

---

Here is a summary of my status for this PR. I believe we only have a few items left.
- [x] Complete run with v3 data
  -  _(note: must run without `arm_diags` due to chunking issue with dataset described [above](https://github.com/E3SM-Project/e3sm_diags/pull/907#issuecomment-2578977046)_
  - [x] Regression testing
     - [Regression testing notebook](https://github.com/E3SM-Project/e3sm_diags/blob/address_diffs_v2.12.1_to_v3/auxiliary_tools/cdat_regression_testing/906-v3_complete_run/regression-test/regression_nc.ipynb)
     - 553/590 match
     - 4/590 nan mismatch (okay, known diff with ccb subsetting)
     - 33/590 not equal (okay, very small number of elements are different with reasonable abs/rel diffs) 
- [ ] Complete run with v2 data --  IN PROGRESS
  - _(note: must run without `arm_diags` due to chunking issue with dataset described [above](https://github.com/E3SM-Project/e3sm_diags/pull/907#issuecomment-2578977046)_
  - [ ] Confirm `convert_units()` update works as intended
  - [ ] Regressing testing
- [ ] Test `arm_diags` with v2 and v3 data with correct chunking 
   - Alternative solution: Call `open_dataset()` when only a single file is needed.

chengzhuzhang commented:
I updated the pre-processing script for single site output, and created correctly chunked datasets: /global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/

The datasets load fast but I ran into an issue with the `diurnal_cycle_zt` subset (error message as follows). I will further investigate if a fix is needed for the newly generated data, or the code base.
```
15:05:14,345 [ERROR]: core_parameter.py(_run_diag:343) >> Error in e3sm_diags.driver.arm_diags_driver
Traceback (most recent call last):
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/edv3/lib/python3.12/site-packages/e3sm_diags/parameter/core_parameter.py", line 340, in _run_diag
    single_result = module.run_diag(self)
                    ^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/edv3/lib/python3.12/site-packages/e3sm_diags/driver/arm_diags_driver.py", line 61, in run_diag
    return _run_diag_diurnal_cycle_zt(parameter)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/edv3/lib/python3.12/site-packages/e3sm_diags/driver/arm_diags_driver.py", line 183, in _run_diag_diurnal_cycle_zt
    test_diurnal, lst = composite_diurnal_cycle(  # type: ignore
                        ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/global/cfs/cdirs/e3sm/zhang40/conda_envs/edv3/lib/python3.12/site-packages/e3sm_diags/driver/utils/diurnal_cycle_xr.py", line 86, in composite_diurnal_cycle
    lst[it, :, ilon] = (itime + start_time + lon[ilon] / 360 * 24) % 24  # type: ignore
    ~~~^^^^^^^^^^^^^
ValueError: could not broadcast input array from shape (131400,) into shape (1,)
```

chengzhuzhang commented:
https://github.com/E3SM-Project/e3sm_diags/pull/907/commits/313a64142717ba60036bccbbabcb885901569f32 should addressed the data problem reported https://github.com/E3SM-Project/e3sm_diags/pull/907#issuecomment-2584735337. The lat, and lon values has time as its dimension which offended the diurnal cycle routine. I'm generating a new set of datasets and will replace the files under `/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/` for testing.

tomvothecoder commented:
> [313a641](https://github.com/E3SM-Project/e3sm_diags/commit/313a64142717ba60036bccbbabcb885901569f32) should addressed the data problem reported [#907 (comment)](https://github.com/E3SM-Project/e3sm_diags/pull/907#issuecomment-2584735337). The lat, and lon values has time as their dimension which offended the diurnal cycle routine. I'm generating a new set of datasets and will replace the files under `/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site/` for testing.
> 
> The old version of site data is now located at `/global/cfs/cdirs/e3sm/chengzhu/tutorial2024/v3.LR.historical_0101/post/atm/site_cdat/`

Thanks Jill. I will do an `arm_diags` only run with v3 data on Tues. 

This PR looks like it's almost done. I'm wrapping up regression testing (v3 is good to go, v2 almost done just `Q` var is off). I'd say we're on track to merge next Wed/Thurs.

chengzhuzhang commented:
I can confirm that the performance issue is resolved by the new dataset in place. The full arm set completed in 2 mins in my test.

chengzhuzhang commented:
> > I checked the metadata for v2 vs v3. The difference is that v2 data has 2 x time steps of v3; v2 has 72 vertical levels and v3 has 80. Is it possible that in the original refactored code, data was loaded in without chunking, while in the new code base, data are loaded as dask arrray and chunked in a weird way?
> 
> There are a few possibilities here (and any combination of them together):
> 
>     1. We did regression testing on v2 data with `open_dataset()` which avoids Dask issues
> 
>     2. We incorporated `open_mfdataset()` and `.load()` later on (e.g., [[Feature]: CDAT Migration: Add support for opening multiple time series datasets #861](https://github.com/E3SM-Project/e3sm_diags/issues/861))
> 
>     3. The v2 data does not have weird chunking like v3 data
> 
> 
To recap on this issue, the root cause is that the original set of site-based output was pre-processed with a script that is based on `ncrcat`, for this specific dataset with var(time,lev) or var(time) dimension, to chunk over time create too many small chunks and caused I/O bottleneck. This problem was not shown up during refactor because, the data set was read in with `open_dataset()` which avoids chunking issue. When we introduced `open_mfdataset()` and `.load()` later, the chunksize is not compatible and caused I/O issue. Now we rewrote the pre-processing script with xarray, which is used to generate correctly chunked time-series files, e.g., chunksize = var dimension. The performance is now good.

tomvothecoder commented:
@chengzhuzhang Great, thank you for summarizing the `arm_diags` I/O issue and confirming the performance with the new e3sm_diags is now good!

I will complete regression testing with v2 data, fix the integration tests, and do a final view before merging this PR.

chengzhuzhang commented:
> I will complete regression testing with v2 data, fix the integration tests, and do a final review before merging this PR.

> @chengzhuzhang Great, thank you for summarizing the `arm_diags` I/O issue and confirming the performance with the new e3sm_diags is now good!
> 
> I will complete regression testing with v2 data, fix the integration tests, and do a final review before merging this PR.

yeah, I'm glad we don't need to update the code base for workaround this data issue. I haven't re-produced v2 site data, so the complete v2 run still has to exclude `arm_diags` set.

tomvothecoder commented on e3sm_diags/driver/qbo_driver.py:
@chengzhuzhang We need to replace `cwt` from scipy with PyWavelets (#912) I'm trying to figure out where to pass the `deg = 6` arg here. If you can try taking a look too that'd be great.

tomvothecoder replied:
Click here for diff: https://github.com/E3SM-Project/e3sm_diags/pull/907/files/2144f51f07a9b855e1b763a3cc5f1a493bac7f68..60f53bc5165bb6bb61b549a3dee6d74627681856#diff-747eb4531216710f7645e357652e33c3fe74c293bbb311a1b9c298a8193f02af

chengzhuzhang replied:
i'm looking at it. `cmor` means totally different here...

chengzhuzhang replied:
All are reasonable! I was thinking along the same line.

xylar replied:
> Constrain `scipy < 1.15` for now -- FYI @xylar for E3SM Unified (will update E3SM Confluence page)

I want to make sure we're on the same page.  This constraint would *not* be in E3SM-Unified, it would be in e3sm_diags.  But it would be important that E3SM-Unified or its other dependencies don't contradict this constraint.

tomvothecoder replied:
> > Constrain `scipy < 1.15` for now -- FYI @xylar for E3SM Unified (will update E3SM Confluence page)
> 
> I want to make sure we're on the same page. This constraint would _not_ be in E3SM-Unified, it would be in e3sm_diags. But it would be important that E3SM-Unified or its other dependencies don't contradict this constraint.

Yup we're on the same page. Thanks for confirming.

chengzhuzhang replied:
@tomvothecoder, see comment from @whannah1 here. We can either implement this as-is, and testing and fine tune (if needed) during testing period. Or pinning `scipy`, and update to use `pywavelets` during testing.

tomvothecoder replied:
@chengzhuzhang I'm going to pin `scipy` here and address #912 in a separate PR.

tomvothecoder commented on auxiliary_tools/cdat_regression_testing/843-migration-phase3/run-script-model-vs-obs/run_script.py:
```diff
@@ -12,7 +12,7 @@
 
 NERSC perlmutter cpu:
-    salloc --nodes 1 --qos interactive --time 01:00:00 --constraint cpu --account=e3sm
+    salloc --nodes 1 --qos interactive --time 0:00:00 --constraint cpu --account=e3sm
```
```suggestion
    salloc --nodes 1 --qos interactive --time 2:00:00 --constraint cpu --account=e3sm
```

tomvothecoder commented on auxiliary_tools/cdat_regression_testing/843-migration-phase3/run-script-model-vs-obs/run_script.py:
```diff
@@ -168,7 +170,7 @@ def run_all_sets():
         "diurnal_cycle",
         "streamflow",
-        "arm_diags",
+        # "arm_diags",
```
```suggestion
        "arm_diags",
```

tomvothecoder commented on auxiliary_tools/cdat_regression_testing/906-v2_complete_run/run_script.py:
```diff
@@ -0,0 +1,289 @@
+    mp_param.test_end_yr = "0060"
+
+    param.save_netcdf = True
+    runner.sets_to_run = [
+        # "lat_lon",
+        # "zonal_mean_xy",
+        # "zonal_mean_2d",
+        # "zonal_mean_2d_stratosphere",
+        # "polar",
+        # "cosp_histogram",
+        # "meridional_mean_2d",
+        # "annual_cycle_zonal_mean",
+        # "enso_diags",
+        # "qbo",
+        # "area_mean_time_series",
+        # "diurnal_cycle",
+        "streamflow",
+        # "arm_diags",
+        # "tc_analysis",
+        # "aerosol_aeronet",
+        # "aerosol_budget",
+        # "mp_partition",
+    ]
```
```suggestion
    runner.sets_to_run = [
        "lat_lon",
        "zonal_mean_xy",
        "zonal_mean_2d",
        "zonal_mean_2d_stratosphere",
        "polar",
        "cosp_histogram",
        "meridional_mean_2d",
        "annual_cycle_zonal_mean",
        "enso_diags",
        "qbo",
        "area_mean_time_series",
        "diurnal_cycle",
        "streamflow",
        "arm_diags",
        "tc_analysis",
        "aerosol_aeronet",
        "aerosol_budget",
        "mp_partition",
    ]
```

tomvothecoder reviewed (commented):
@chengzhuzhang Here is my final self-review. All of the code changes look good to me. I will merge if you agree with the testing results below.

## v2 data complete run results [(notebook)](https://github.com/E3SM-Project/e3sm_diags/blob/address_diffs_v2.12.1_to_v3/auxiliary_tools/cdat_regression_testing/906-v2_complete_run/regression_nc.ipynb):

Results:

We are good to go here. The differences are expected and explained in the comments below.

```python
                stat_name  value       pct
0    matching_files_count   1093  0.875801
1      missing_vars_count      0  0.000000
2   mismatch_errors_count     10  0.008013
3  not_equal_errors_count    143  0.114583
4        key_errors_count      0  0.000000
5     missing_files_count      2  0.001603
```

- 1093/1246 matching files
- 10/1246 mismatch errors due to ccb regional subsetting differences
- 143/1246 not equal files -- number of different elements are really small and stats (min, max, mean, sum) are similar, most likely related to the `convert_units()` change to replace `genutil.udunits` with `cf_units.Unit`.
  - Except `Q` variable, differences are explained by not updating the `Q` results for main branch with unit conversion changes. Non-issue.
- 2/1246 are missing because `AOD_550` retired.


## v3 data complete run results ([notebook](https://github.com/E3SM-Project/e3sm_diags/blob/address_diffs_v2.12.1_to_v3/auxiliary_tools/cdat_regression_testing/906-v3_complete_run/regression_nc.ipynb))

We are good to go here. The differences are expected and explained in the comments below.

```python
                stat_name  value       pct
0    matching_files_count    553  0.937288
1      missing_vars_count      0  0.000000
2   mismatch_errors_count      4  0.006780
3  not_equal_errors_count     33  0.055932
4        key_errors_count      0  0.000000
5     missing_files_count      0  0.000000
```

- 553 match within the rtol of 1e-5, which is awesome.
- 4 mismatch errors are known issues due to regional subsetting differences with "ccb" flag
- 33 not equal errors are not a concern because they affect very small number of elements in the dataset. The stats (min, max, mean, and sum) of the datasets between branches are close.

chengzhuzhang commented:
@tomvothecoder thank you for the regression tests! the metrics data look good for releasing the first `rc`. To not to delay the testing period of e3sm-unified, I suggest to merge. I will go through the diffs and we can resolve remaining issues, if needed, during testing period.

chengzhuzhang commented:
@tomvothecoder somehow, I can't find `25-01-15-branch-907` under `https://portal.nersc.gov/cfs/e3sm/e3sm_diags/complete_run/`

tomvothecoder commented:
> @tomvothecoder thank you for the regression tests! the metrics data look good for releasing the first `rc`. To not to delay the testing period of e3sm-unified, I suggest to merge. I will go through the diffs and we can resolve remaining issues, if needed, during testing period.

Sounds good!

> @tomvothecoder somehow, I can't find `25-01-15-branch-907` under `https://portal.nersc.gov/cfs/e3sm/e3sm_diags/complete_run/`

I forgot to make some of the directories `o=rx`. I will do that once perlmutter is back online.

tomvothecoder commented:
> @tomvothecoder somehow, I can't find `25-01-15-branch-907` under `https://portal.nersc.gov/cfs/e3sm/e3sm_diags/complete_run/`

The directory is now available here: https://portal.nersc.gov/cfs/e3sm/e3sm_diags/complete_run/25-01-15-branch-907/
